### PR TITLE
cache visibility check results for reuse within a single operation

### PR DESCRIPTION
**Pull request by just-boris**
My naive attempt to fix: theKashey/react-focus-lock#151

I confirmed this in my real application case, it prevents Jest test from getting timed out.

### CODE REVIEW

**Comment by theKashey** [comment]
> size-limit is failing the build. Is configured a bit tight.

Thank you for making this change and proving that it is a fix. However, before we proceed  I need to do one more little thing - I want to track the existing "complexity" to understand the difference.

> PS: technically speaking there is a slightly better way to manage this use case - wrap the single memoization target - `isVisible` with [kashe](https://github.com/theKashey/kashe), and _switch universes_ at the top (`visibilityCache`) to throw-away ones using `inboxed` helper.
Unfortunately, that is +1kb of the code, which is +30% to the total size of this library.

**Comment by theKashey** [comment]
Complexity check has been merged - https://github.com/theKashey/focus-lock/pull/25 - and everything is VERY BAD

Please merge with a master and lets see much much crap will be gone.

File: src/utils/is.ts [hunk 1]
```diff
@@ -7,17 +7,28 @@ const isElementHidden = (computedStyle: CSSStyleDeclaration): boolean => {
...
-      node.parentNode && node.parentNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE
-        ? (node.parentNode as any).host
-        : node.parentNode
-    ));
+export const isVisible = (node: HTMLElement | undefined, visibilityCache: Map<HTMLElement, boolean>): boolean => {
+    if(!node) {
+        return false;
+    }
+    const cached = visibilityCache.get(node)
+    if(cached) {
```
Comment (theKashey) [inline src/utils/is.ts:15]: that is working only for "true" cases, while false is legit as well
```suggestion
    if (cached!==undefined) {
```

File: src/utils/is.ts [hunk 2]
```diff
@@ -7,17 +7,28 @@ const isElementHidden = (computedStyle: CSSStyleDeclaration): boolean => {
...
+        return cached;
+    }
+    // @ts-ignore
+    const result = node === document ||
```
Comment (theKashey) [inline src/utils/is.ts:19]: can we extract the actual check to the separate function?
Technically you can just rename "the old one" and then build a cached version next to it.

**Review by theKashey — CHANGES_REQUESTED** [review]
- one suggestion
- and one new test to pass

**Review by theKashey — APPROVED** [review]

**Comment by just-boris** [comment]
ok, so the fix got merged, but to benefit from it we need to have a release

Is there anything blocking this?

**Comment by theKashey** [comment]
I was just not satisfied with the result numbers and added a few more tests.
As a result I've found that with the very last commit you broke nested memoization - it was caching result only for the very first noded. Not everything.

The result numbers are WAY better, and (again) that's all thanks to you 🎉

File: src/utils/is.ts [hunk 3]
```diff
@@ -7,7 +7,7 @@ const isElementHidden = (computedStyle: CSSStyleDeclaration): boolean => {
   );
 };
 
-export const isVisible = (node: HTMLElement | undefined): boolean =>
+const isVisible = (node: HTMLElement | undefined): boolean =>
   !node ||
   // @ts-ignore
   node === document ||
```
Comment (theKashey) [inline src/utils/is.ts:13]: oh no. Nested call is to the same "uncached" isVisible

**Comment by theKashey** [comment]
Published as focus-lock@0.9.1

**Comment by just-boris** [comment]
Very nice! Thanks for doing proper quality assurance of my code :)

**Pull request merged**